compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -149,12 +149,36 @@ object Inliner {
 
   /** Replace `Inlined` node by a block that contains its bindings and expansion */
   def dropInlined(inlined: tpd.Inlined)(implicit ctx: Context): Tree = {
-    val reposition = new TreeMap {
-      override def transform(tree: Tree)(implicit ctx: Context): Tree = {
-        super.transform(tree).withPos(inlined.call.pos)
+    if (enclosingInlineds.nonEmpty) inlined // Remove in the outer most inlined call
+    else {
+      // Position used for any tree that was inlined (including recursive inlines)
+      val inlinedAtPos = inlined.call.pos
+
+      /** Removes all Inlined trees, replacing them with blocks.
+       *  Repositions all trees directly inside an inlined expansion of a non empty call to the position of the call.
+       *  Any tree directly inside an empty call (inlined in the inlined code) retains their position.
+       */
+      class Reposition extends TreeMap {
+        override def transform(tree: Tree)(implicit ctx: Context): Tree = {
+          tree match {
+            case tree: Inlined => transformInline(tree)
+            case _ =>
+              val transformed = super.transform(tree)
+              enclosingInlineds match {
+                case call :: _ if call.symbol.sourceFile != ctx.owner.sourceFile =>
+                  // reposition tree inlined from some other file
```

Comment:
I think the documentation is too complicated. We're dropping Inlined nodes, and repositioning nodes from other files to the call position because we can't express their position otherwise in their debug info, pending JSR-45. The stuff about "empty call" is confusing and expressed in terms of the implementation.

Other issues that I think are there and that I'm trying to address:

- The comment makes sense, but the code is stricter, and resets the position even if the toplevel class is different but the source file is the same. In that case, it appears we don't need to reset the position. Addressed by one commit I pushed.

- Also, this is an expensive (O(n)) way to test if enclosingInlineds.nonEmpty and get the head, and that could be expensive, but that's really enclosingInlineds's fault and I think can be fixed separately without affecting this code.